app/sidekiq/kms_key_rotation/batch_initiator_job.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module KmsKeyRotation
+  class BatchComplete
+    def on_complete(_status, _options)
+      BatchInitiatorJob.perform_async
+    end
+  end
+
+  class BatchInitiatorJob
+    include Sidekiq::Worker
+
+    sidekiq_options retry: 5, queue: :low
+
+    RECORDS_PER_BATCH = 10_000
+    RECORDS_PER_JOB = 1_000
+
+    def perform
+      batch = Sidekiq::Batch.new
+      batch.description = "KMS Key Rotation #{batch.bid}"
+      batch.on(:complete, BatchComplete)
+
+      return nil if records.empty?
+
+      batch.jobs do
+        Sidekiq::Client.push_bulk('class' => KmsKeyRotation::RotateKeysJob, 'args' => batched_gids)
+      end
+    end
+
+    def records
+      @records ||= get_records
+    end
+
+    private
+
+    def get_records
+      records = []
+      get_models.each do |model|
+        break if records.size == RECORDS_PER_BATCH
+
+        records += model.where.not('encrypted_kms_key LIKE ? ', "v#{Time.zone.today.year}:%")
+                        .limit(RECORDS_PER_BATCH - records.size)
+      end
+      records
+    end
+
+    def get_models
+      ApplicationRecord.descendants_using_encryption.map(&:name).map(&:constantize)
+    end
+
+    def batched_gids
+      batched_gids = []
+
+      begin
+        records.map(&:to_global_id)
+               .each_slice(RECORDS_PER_JOB) do |gids|
+                 batched_gids << [{ gids: }.to_json]
+               end
+      rescue => e
+        Rails.logger.error("Error batching jobs: #{e.message}")
+      end
```

Comment:
Do we want to capture the errors and let it continue or should we re-raise here?

Reply:
so I guess my original thought was that it would log the error, batched_gids would be empty, and the job would then end. But yes, besides the error message in the logs, nothing would tell the system that the job failed and stopped propagating... I'll add something